cpp/src/slice2cs/Gen.cpp

```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
+                                   false,
+                                   ns,
+                                   members.size() == 1 ? "value" : "value." + fieldName(member), member->tag(),
+                                   stream);
+        }
+
+        if (repeat)
+        {
+            repeat = false;
+        }
+        else if (write11ReturnLast)
+        {
+            _out << eb;
+            _out << nl;
+            _out << "else";
+            _out << sb;
+
+            // Repeat after updating requiredMembers
+            repeat = true;
+            requiredMembers.push_back(requiredMembers.front());
+            requiredMembers.pop_front();
+        }
     }
+    while (repeat)
+        ;
 
-    for (const auto& param : taggedParams)
+    if (write11ReturnLast)
     {
-        writeTaggedMarshalCode(_out,
-                               OptionalPtr::dynamicCast(param->type()),
-                               false,
-                               ns,
-                               singleParam ? "value" : obj + fieldName(param), param->tag(),
-                               stream);
+        _out << eb;
     }
 }
 
 void
-Slice::CsVisitor::writeUnmarshalParams(const OperationPtr& op,
-                                       const MemberList& requiredParams,
-                                       const MemberList& taggedParams,
-                                       bool isReturnValue,
-                                       const string& customStream)
+Slice::CsVisitor::writeUnmarshal(const OperationPtr& operation, bool returnType)
 {
-    const string stream = customStream.empty() ? "istr" : customStream;
-    string ns = getNamespace(op->interface());
+    const string stream = "istr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
-    if (requiredParams.size() > 0)
+    if (requiredMembers.size() > 0)
     {
-        if (isReturnValue)
+        if (returnType)
         {
-            bitSequenceSize = op->returnBitSequenceSize();
+            bitSequenceSize = operation->returnBitSequenceSize();
         }
         else
         {
-            bitSequenceSize = op->paramsBitSequenceSize();
+            bitSequenceSize = operation->paramsBitSequenceSize();
         }
+    }
 
+    bool read11ReturnLast = returnType && operation->hasReturnAndOut() && requiredMembers.size() > 1 &&
+        !members.front()->tagged();
+
+    if (read11ReturnLast)
+    {
+        _out << nl << "if (" << stream << ".Encoding != ZeroC.Ice.Encoding.V1_1)";
+        _out << sb;
+    }
+
+    bool repeat = false;
+
+    do
+    {
         if (bitSequenceSize > 0)
         {
             _out << nl << "var bitSequence = " << stream << ".ReadBitSequence(" << bitSequenceSize << ");";
```

Comment:
same here we can avoid declaring bitSequence twice

Reply:
Fixed a suggested.